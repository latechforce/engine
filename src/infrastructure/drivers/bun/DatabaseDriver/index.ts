import type { IDatabaseDriver } from '/adapter/spi/drivers/DatabaseSpi'
import type { DatabaseConfig, DatabaseEventType } from '/domain/services/Database'
import type { EventDto } from '/adapter/spi/dtos/EventDto'
import { PostgreSQLDatabaseDriver } from '/infrastructure/drivers/common/DatabaseDriver/PostgreSQLDriver'
import { SQLiteDatabaseDriver } from './SQLite/SQLiteDriver'
import type { AutomationHistoryRecord } from '/domain/entities/Automation/History'
import type { TableConfig } from '/domain/entities/Table'
import type { TableSchema } from '../../../../adapter/api/schemas/TableSchema'
import { TableMapper } from '/adapter/api/mappers/TableMapper'

export class DatabaseDriver implements IDatabaseDriver {
  private _db: PostgreSQLDatabaseDriver | SQLiteDatabaseDriver
  public type: DatabaseConfig['type']

  constructor(config: DatabaseConfig) {
    this.type = config.type
    switch (this.type) {
      case 'SQLite':
        this._db = new SQLiteDatabaseDriver(config)
        break
      case 'PostgreSQL':
        this._db = new PostgreSQLDatabaseDriver(config)
        break
      default:
        throw new Error('Invalid driver')
    }
  }

  connect = async (): Promise<void> => {
    await this._db.connect()
  }

  disconnect = async (): Promise<void> => {
    await this._db.disconnect()
  }

  exec = async (query: string): Promise<void> => {
    await this._db.exec(query)
  }

  query = async <T>(
    text: string,
    values: (string | number | Buffer | Date)[] = []
  ): Promise<{ rows: T[]; rowCount: number }> => {
    return this._db.query(text, values)
  }

  table(table: TableConfig) {
    return this._db.table(table)
  }

  tableFromSchema(schema: TableSchema) {
    return this._db.table(TableMapper.toConfig(schema))
  }

  on = (event: DatabaseEventType, callback: (eventDto: EventDto) => void) => {
    this._db.on(event, callback)
  }

  setupTriggers = async (tables: string[]) => {
    await this._db.setupTriggers(tables)
  }

  waitForAutomationsHistories = async ({
    count = 1,
    status = 'succeed',
    timeout = 60000,
  }: {
    count?: number
    status?: string
    timeout?: number
  } = {}) => {
    let histories: AutomationHistoryRecord[] = []
    let retry = 0
    do {
      await new Promise((resolve) => setTimeout(resolve, 500))
      const { rows } = await this.query<AutomationHistoryRecord>(
        'SELECT * FROM automations_histories_view WHERE status = ?',
        [status]
      )
      histories = rows
      retry++
    } while (histories.length < count && retry < Math.floor(timeout / 500))
    if (histories.length < count) {
      throw new Error(`Automations histories not found: ${count} ${status} histories`)
    }
    return histories
  }
}
