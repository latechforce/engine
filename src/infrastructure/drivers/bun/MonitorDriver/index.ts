import type { MonitorConfig } from '/domain/services/Monitor'
import type { IMonitorDriver } from '/adapter/spi/drivers/MonitorSpi'
import { SentryDriver } from './SentryDriver'
import { ConsoleDriver } from '../../common/MonitorDriver/ConsoleDriver'

export class MonitorDriver implements IMonitorDriver {
  private _monitors: (SentryDriver | ConsoleDriver)[] = []

  constructor(config: MonitorConfig[]) {
    for (const monitor of config) {
      const { driver } = monitor
      switch (driver) {
        case 'Sentry':
          this._monitors.push(new SentryDriver())
          break
        case 'Console':
          this._monitors.push(new ConsoleDriver())
          break
        default:
          throw new Error('Invalid driver')
      }
    }
  }

  captureException = (error: Error) => {
    this._monitors.map((m) => m.captureException(error))
  }

  captureMessage = (message: string) => {
    this._monitors.map((m) => m.captureMessage(message))
  }
}
