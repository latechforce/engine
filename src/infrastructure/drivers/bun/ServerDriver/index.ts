import type { IServerDriver } from '/adapter/spi/drivers/ServerSpi'
import type { DeleteDto, GetDto, PatchDto, PostDto, RequestDto } from '/adapter/spi/dtos/RequestDto'
import type { ServerConfig, ServerMethodOptions } from '/domain/services/Server'
import type { Response } from '/domain/entities/Response'
import { ElysiaDriver } from './ElysiaDriver'

export class ServerDriver implements IServerDriver {
  private _server: ElysiaDriver

  constructor(config: ServerConfig) {
    this._server = new ElysiaDriver(config)
  }

  get = async (
    path: string,
    handler: (getDto: GetDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.get(path, handler, options)
  }

  post = async (
    path: string,
    handler: (postDto: PostDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.post(path, handler, options)
  }

  patch = async (
    path: string,
    handler: (patchDto: PatchDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.patch(path, handler, options)
  }

  delete = async (
    path: string,
    handler: (deleteDto: DeleteDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.delete(path, handler, options)
  }

  notFound = async (handler: (requestDto: RequestDto) => Promise<Response>) => {
    await this._server.notFound(handler)
  }

  findAvailablePort = async (): Promise<number> => {
    return this._server.findAvailablePort()
  }

  start = async (port: number): Promise<number> => {
    return this._server.start(port)
  }

  stop = async () => {
    await this._server.stop()
  }
}
