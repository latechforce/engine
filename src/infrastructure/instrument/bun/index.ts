import type { ConfigSchema } from '/adapter/api/schemas/ConfigSchema'
import * as Sentry from './Sentry'
import { SystemDriver } from '/infrastructure/drivers/common/SystemDriver'

export function instrument(config: ConfigSchema) {
  const system = new SystemDriver()
  if (config.services?.monitors) {
    for (const monitor of config.services.monitors) {
      if (monitor.driver === 'Sentry')
        Sentry.init({
          appName: config.name,
          appVersion: config.appVersion ?? system.getAppVersion(),
          ...monitor,
        })
    }
  }
}
