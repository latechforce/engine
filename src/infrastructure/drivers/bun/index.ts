import type { Drivers } from '/adapter/spi/drivers'
import type { DatabaseConfig } from '/domain/services/Database'
import type { MonitorsConfig } from '/domain/services/Monitor'
import type { ServerConfig } from '/domain/services/Server'
import type { StorageConfig } from '/domain/services/Storage'

import { DatabaseDriver } from './DatabaseDriver'
import { MonitorDriver } from './MonitorDriver'
import { ServerDriver } from './ServerDriver'

import { drivers as common } from '../common'
import { StorageDriver } from './StorageDriver'

export const drivers: Drivers = {
  ...common,
  monitor: (config: MonitorsConfig) => new MonitorDriver(config),
  database: (config: DatabaseConfig) => new DatabaseDriver(config),
  server: (config: ServerConfig) => new ServerDriver(config),
  storage: (config: StorageConfig) => new StorageDriver(config),
}
