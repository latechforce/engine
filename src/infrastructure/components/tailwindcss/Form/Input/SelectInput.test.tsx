import { it, expect } from 'bun:test'
import { SelectInput } from './SelectInput'
import { renderToString } from 'react-dom/server'

it('should match the select input snapshot', async () => {
  // GIVEN
  const input = <SelectInput field="name" options={['Option 1', 'Option 2', 'Option 3']} />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the select input snapshot with label', async () => {
  // GIVEN
  const input = (
    <SelectInput field="name" label="Name" options={['Option 1', 'Option 2', 'Option 3']} />
  )

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the select input snapshot with description', async () => {
  // GIVEN
  const input = (
    <SelectInput
      field="name"
      description="Description"
      options={['Option 1', 'Option 2', 'Option 3']}
    />
  )

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the select input snapshot with label and description', async () => {
  // GIVEN
  const input = (
    <SelectInput
      field="name"
      label="Name"
      description="Description"
      options={['Option 1', 'Option 2', 'Option 3']}
    />
  )

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the select input snapshot with placeholder', async () => {
  // GIVEN
  const input = (
    <SelectInput
      field="name"
      placeholder="Placeholder"
      options={['Option 1', 'Option 2', 'Option 3']}
    />
  )

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the select input snapshot with required', async () => {
  // GIVEN
  const input = <SelectInput field="name" required options={['Option 1', 'Option 2', 'Option 3']} />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})
