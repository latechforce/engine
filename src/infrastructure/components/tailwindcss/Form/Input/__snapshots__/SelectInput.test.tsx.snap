// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the select input snapshot 1`] = `
"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name</label><select name="name" data-hs-select="{
        &quot;placeholder&quot;: &quot;&quot;,
        &quot;toggleTag&quot;: &quot;&lt;button type=\\&quot;button\\&quot; aria-expanded=\\&quot;false\\&quot;&gt;&lt;/button&gt;&quot;,
        &quot;toggleClasses&quot;: &quot;hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 relative py-3 ps-4 pe-9 flex gap-x-2 text-nowrap w-full cursor-pointer bg-white border border-gray-200 rounded-lg text-start text-sm focus:outline-hidden focus:ring-2 focus:ring-blue-500 dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:focus:outline-hidden dark:focus:ring-1 dark:focus:ring-neutral-600&quot;,
        &quot;dropdownClasses&quot;: &quot;mt-2 z-50 w-full max-h-72 p-1 space-y-0.5 bg-white border border-gray-200 rounded-lg overflow-hidden overflow-y-auto [&amp;::-webkit-scrollbar]:w-2 [&amp;::-webkit-scrollbar-thumb]:rounded-full [&amp;::-webkit-scrollbar-track]:bg-gray-100 [&amp;::-webkit-scrollbar-thumb]:bg-gray-300 dark:[&amp;::-webkit-scrollbar-track]:bg-neutral-700 dark:[&amp;::-webkit-scrollbar-thumb]:bg-neutral-500 dark:bg-neutral-900 dark:border-neutral-700&quot;,
        &quot;optionClasses&quot;: &quot;py-2 px-4 w-full text-sm text-gray-800 cursor-pointer hover:bg-gray-100 rounded-lg focus:outline-hidden focus:bg-gray-100 hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:text-neutral-200 dark:focus:bg-neutral-800&quot;,
        &quot;optionTemplate&quot;: &quot;&lt;div class=\\&quot;flex justify-between items-center w-full\\&quot;&gt;&lt;span data-title&gt;&lt;/span&gt;&lt;span class=\\&quot;hidden hs-selected:block\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-blue-600 dark:text-blue-500 \\&quot; xmlns=\\&quot;http:.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;polyline points=\\&quot;20 6 9 17 4 12\\&quot;/&gt;&lt;/svg&gt;&lt;/span&gt;&lt;/div&gt;&quot;,
        &quot;extraMarkup&quot;: &quot;&lt;div class=\\&quot;absolute top-1/2 end-3 -translate-y-1/2\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-gray-500 dark:text-neutral-500 \\&quot; xmlns=\\&quot;http://www.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;path d=\\&quot;m7 15 5 5 5-5\\&quot;/&gt;&lt;path d=\\&quot;m7 9 5-5 5 5\\&quot;/&gt;&lt;/svg&gt;&lt;/div&gt;&quot;
      }" class="hidden"><option value=""></option><option>Option 1</option><option>Option 2</option><option>Option 3</option></select></div>"
`;

exports[`should match the select input snapshot with label 1`] = `
"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">Name</label><select name="name" data-hs-select="{
        &quot;placeholder&quot;: &quot;&quot;,
        &quot;toggleTag&quot;: &quot;&lt;button type=\\&quot;button\\&quot; aria-expanded=\\&quot;false\\&quot;&gt;&lt;/button&gt;&quot;,
        &quot;toggleClasses&quot;: &quot;hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 relative py-3 ps-4 pe-9 flex gap-x-2 text-nowrap w-full cursor-pointer bg-white border border-gray-200 rounded-lg text-start text-sm focus:outline-hidden focus:ring-2 focus:ring-blue-500 dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:focus:outline-hidden dark:focus:ring-1 dark:focus:ring-neutral-600&quot;,
        &quot;dropdownClasses&quot;: &quot;mt-2 z-50 w-full max-h-72 p-1 space-y-0.5 bg-white border border-gray-200 rounded-lg overflow-hidden overflow-y-auto [&amp;::-webkit-scrollbar]:w-2 [&amp;::-webkit-scrollbar-thumb]:rounded-full [&amp;::-webkit-scrollbar-track]:bg-gray-100 [&amp;::-webkit-scrollbar-thumb]:bg-gray-300 dark:[&amp;::-webkit-scrollbar-track]:bg-neutral-700 dark:[&amp;::-webkit-scrollbar-thumb]:bg-neutral-500 dark:bg-neutral-900 dark:border-neutral-700&quot;,
        &quot;optionClasses&quot;: &quot;py-2 px-4 w-full text-sm text-gray-800 cursor-pointer hover:bg-gray-100 rounded-lg focus:outline-hidden focus:bg-gray-100 hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:text-neutral-200 dark:focus:bg-neutral-800&quot;,
        &quot;optionTemplate&quot;: &quot;&lt;div class=\\&quot;flex justify-between items-center w-full\\&quot;&gt;&lt;span data-title&gt;&lt;/span&gt;&lt;span class=\\&quot;hidden hs-selected:block\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-blue-600 dark:text-blue-500 \\&quot; xmlns=\\&quot;http:.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;polyline points=\\&quot;20 6 9 17 4 12\\&quot;/&gt;&lt;/svg&gt;&lt;/span&gt;&lt;/div&gt;&quot;,
        &quot;extraMarkup&quot;: &quot;&lt;div class=\\&quot;absolute top-1/2 end-3 -translate-y-1/2\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-gray-500 dark:text-neutral-500 \\&quot; xmlns=\\&quot;http://www.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;path d=\\&quot;m7 15 5 5 5-5\\&quot;/&gt;&lt;path d=\\&quot;m7 9 5-5 5 5\\&quot;/&gt;&lt;/svg&gt;&lt;/div&gt;&quot;
      }" class="hidden"><option value=""></option><option>Option 1</option><option>Option 2</option><option>Option 3</option></select></div>"
`;

exports[`should match the select input snapshot with description 1`] = `
"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name</label><p class="mb-2 text-sm text-gray-500 dark:text-neutral-500">Description</p><select name="name" data-hs-select="{
        &quot;placeholder&quot;: &quot;&quot;,
        &quot;toggleTag&quot;: &quot;&lt;button type=\\&quot;button\\&quot; aria-expanded=\\&quot;false\\&quot;&gt;&lt;/button&gt;&quot;,
        &quot;toggleClasses&quot;: &quot;hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 relative py-3 ps-4 pe-9 flex gap-x-2 text-nowrap w-full cursor-pointer bg-white border border-gray-200 rounded-lg text-start text-sm focus:outline-hidden focus:ring-2 focus:ring-blue-500 dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:focus:outline-hidden dark:focus:ring-1 dark:focus:ring-neutral-600&quot;,
        &quot;dropdownClasses&quot;: &quot;mt-2 z-50 w-full max-h-72 p-1 space-y-0.5 bg-white border border-gray-200 rounded-lg overflow-hidden overflow-y-auto [&amp;::-webkit-scrollbar]:w-2 [&amp;::-webkit-scrollbar-thumb]:rounded-full [&amp;::-webkit-scrollbar-track]:bg-gray-100 [&amp;::-webkit-scrollbar-thumb]:bg-gray-300 dark:[&amp;::-webkit-scrollbar-track]:bg-neutral-700 dark:[&amp;::-webkit-scrollbar-thumb]:bg-neutral-500 dark:bg-neutral-900 dark:border-neutral-700&quot;,
        &quot;optionClasses&quot;: &quot;py-2 px-4 w-full text-sm text-gray-800 cursor-pointer hover:bg-gray-100 rounded-lg focus:outline-hidden focus:bg-gray-100 hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:text-neutral-200 dark:focus:bg-neutral-800&quot;,
        &quot;optionTemplate&quot;: &quot;&lt;div class=\\&quot;flex justify-between items-center w-full\\&quot;&gt;&lt;span data-title&gt;&lt;/span&gt;&lt;span class=\\&quot;hidden hs-selected:block\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-blue-600 dark:text-blue-500 \\&quot; xmlns=\\&quot;http:.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;polyline points=\\&quot;20 6 9 17 4 12\\&quot;/&gt;&lt;/svg&gt;&lt;/span&gt;&lt;/div&gt;&quot;,
        &quot;extraMarkup&quot;: &quot;&lt;div class=\\&quot;absolute top-1/2 end-3 -translate-y-1/2\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-gray-500 dark:text-neutral-500 \\&quot; xmlns=\\&quot;http://www.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;path d=\\&quot;m7 15 5 5 5-5\\&quot;/&gt;&lt;path d=\\&quot;m7 9 5-5 5 5\\&quot;/&gt;&lt;/svg&gt;&lt;/div&gt;&quot;
      }" class="hidden"><option value=""></option><option>Option 1</option><option>Option 2</option><option>Option 3</option></select></div>"
`;

exports[`should match the select input snapshot with label and description 1`] = `
"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">Name</label><p class="mb-2 text-sm text-gray-500 dark:text-neutral-500">Description</p><select name="name" data-hs-select="{
        &quot;placeholder&quot;: &quot;&quot;,
        &quot;toggleTag&quot;: &quot;&lt;button type=\\&quot;button\\&quot; aria-expanded=\\&quot;false\\&quot;&gt;&lt;/button&gt;&quot;,
        &quot;toggleClasses&quot;: &quot;hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 relative py-3 ps-4 pe-9 flex gap-x-2 text-nowrap w-full cursor-pointer bg-white border border-gray-200 rounded-lg text-start text-sm focus:outline-hidden focus:ring-2 focus:ring-blue-500 dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:focus:outline-hidden dark:focus:ring-1 dark:focus:ring-neutral-600&quot;,
        &quot;dropdownClasses&quot;: &quot;mt-2 z-50 w-full max-h-72 p-1 space-y-0.5 bg-white border border-gray-200 rounded-lg overflow-hidden overflow-y-auto [&amp;::-webkit-scrollbar]:w-2 [&amp;::-webkit-scrollbar-thumb]:rounded-full [&amp;::-webkit-scrollbar-track]:bg-gray-100 [&amp;::-webkit-scrollbar-thumb]:bg-gray-300 dark:[&amp;::-webkit-scrollbar-track]:bg-neutral-700 dark:[&amp;::-webkit-scrollbar-thumb]:bg-neutral-500 dark:bg-neutral-900 dark:border-neutral-700&quot;,
        &quot;optionClasses&quot;: &quot;py-2 px-4 w-full text-sm text-gray-800 cursor-pointer hover:bg-gray-100 rounded-lg focus:outline-hidden focus:bg-gray-100 hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:text-neutral-200 dark:focus:bg-neutral-800&quot;,
        &quot;optionTemplate&quot;: &quot;&lt;div class=\\&quot;flex justify-between items-center w-full\\&quot;&gt;&lt;span data-title&gt;&lt;/span&gt;&lt;span class=\\&quot;hidden hs-selected:block\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-blue-600 dark:text-blue-500 \\&quot; xmlns=\\&quot;http:.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;polyline points=\\&quot;20 6 9 17 4 12\\&quot;/&gt;&lt;/svg&gt;&lt;/span&gt;&lt;/div&gt;&quot;,
        &quot;extraMarkup&quot;: &quot;&lt;div class=\\&quot;absolute top-1/2 end-3 -translate-y-1/2\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-gray-500 dark:text-neutral-500 \\&quot; xmlns=\\&quot;http://www.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;path d=\\&quot;m7 15 5 5 5-5\\&quot;/&gt;&lt;path d=\\&quot;m7 9 5-5 5 5\\&quot;/&gt;&lt;/svg&gt;&lt;/div&gt;&quot;
      }" class="hidden"><option value=""></option><option>Option 1</option><option>Option 2</option><option>Option 3</option></select></div>"
`;

exports[`should match the select input snapshot with placeholder 1`] = `
"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name</label><select name="name" data-hs-select="{
        &quot;placeholder&quot;: &quot;Placeholder&quot;,
        &quot;toggleTag&quot;: &quot;&lt;button type=\\&quot;button\\&quot; aria-expanded=\\&quot;false\\&quot;&gt;&lt;/button&gt;&quot;,
        &quot;toggleClasses&quot;: &quot;hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 relative py-3 ps-4 pe-9 flex gap-x-2 text-nowrap w-full cursor-pointer bg-white border border-gray-200 rounded-lg text-start text-sm focus:outline-hidden focus:ring-2 focus:ring-blue-500 dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:focus:outline-hidden dark:focus:ring-1 dark:focus:ring-neutral-600&quot;,
        &quot;dropdownClasses&quot;: &quot;mt-2 z-50 w-full max-h-72 p-1 space-y-0.5 bg-white border border-gray-200 rounded-lg overflow-hidden overflow-y-auto [&amp;::-webkit-scrollbar]:w-2 [&amp;::-webkit-scrollbar-thumb]:rounded-full [&amp;::-webkit-scrollbar-track]:bg-gray-100 [&amp;::-webkit-scrollbar-thumb]:bg-gray-300 dark:[&amp;::-webkit-scrollbar-track]:bg-neutral-700 dark:[&amp;::-webkit-scrollbar-thumb]:bg-neutral-500 dark:bg-neutral-900 dark:border-neutral-700&quot;,
        &quot;optionClasses&quot;: &quot;py-2 px-4 w-full text-sm text-gray-800 cursor-pointer hover:bg-gray-100 rounded-lg focus:outline-hidden focus:bg-gray-100 hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:text-neutral-200 dark:focus:bg-neutral-800&quot;,
        &quot;optionTemplate&quot;: &quot;&lt;div class=\\&quot;flex justify-between items-center w-full\\&quot;&gt;&lt;span data-title&gt;&lt;/span&gt;&lt;span class=\\&quot;hidden hs-selected:block\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-blue-600 dark:text-blue-500 \\&quot; xmlns=\\&quot;http:.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;polyline points=\\&quot;20 6 9 17 4 12\\&quot;/&gt;&lt;/svg&gt;&lt;/span&gt;&lt;/div&gt;&quot;,
        &quot;extraMarkup&quot;: &quot;&lt;div class=\\&quot;absolute top-1/2 end-3 -translate-y-1/2\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-gray-500 dark:text-neutral-500 \\&quot; xmlns=\\&quot;http://www.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;path d=\\&quot;m7 15 5 5 5-5\\&quot;/&gt;&lt;path d=\\&quot;m7 9 5-5 5 5\\&quot;/&gt;&lt;/svg&gt;&lt;/div&gt;&quot;
      }" class="hidden"><option value=""></option><option>Option 1</option><option>Option 2</option><option>Option 3</option></select></div>"
`;

exports[`should match the select input snapshot with required 1`] = `
"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name<span class="text-red-500"> *</span></label><select name="name" required="" data-hs-select="{
        &quot;placeholder&quot;: &quot;&quot;,
        &quot;toggleTag&quot;: &quot;&lt;button type=\\&quot;button\\&quot; aria-expanded=\\&quot;false\\&quot;&gt;&lt;/button&gt;&quot;,
        &quot;toggleClasses&quot;: &quot;hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 relative py-3 ps-4 pe-9 flex gap-x-2 text-nowrap w-full cursor-pointer bg-white border border-gray-200 rounded-lg text-start text-sm focus:outline-hidden focus:ring-2 focus:ring-blue-500 dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:focus:outline-hidden dark:focus:ring-1 dark:focus:ring-neutral-600&quot;,
        &quot;dropdownClasses&quot;: &quot;mt-2 z-50 w-full max-h-72 p-1 space-y-0.5 bg-white border border-gray-200 rounded-lg overflow-hidden overflow-y-auto [&amp;::-webkit-scrollbar]:w-2 [&amp;::-webkit-scrollbar-thumb]:rounded-full [&amp;::-webkit-scrollbar-track]:bg-gray-100 [&amp;::-webkit-scrollbar-thumb]:bg-gray-300 dark:[&amp;::-webkit-scrollbar-track]:bg-neutral-700 dark:[&amp;::-webkit-scrollbar-thumb]:bg-neutral-500 dark:bg-neutral-900 dark:border-neutral-700&quot;,
        &quot;optionClasses&quot;: &quot;py-2 px-4 w-full text-sm text-gray-800 cursor-pointer hover:bg-gray-100 rounded-lg focus:outline-hidden focus:bg-gray-100 hs-select-disabled:pointer-events-none hs-select-disabled:opacity-50 dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:text-neutral-200 dark:focus:bg-neutral-800&quot;,
        &quot;optionTemplate&quot;: &quot;&lt;div class=\\&quot;flex justify-between items-center w-full\\&quot;&gt;&lt;span data-title&gt;&lt;/span&gt;&lt;span class=\\&quot;hidden hs-selected:block\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-blue-600 dark:text-blue-500 \\&quot; xmlns=\\&quot;http:.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;polyline points=\\&quot;20 6 9 17 4 12\\&quot;/&gt;&lt;/svg&gt;&lt;/span&gt;&lt;/div&gt;&quot;,
        &quot;extraMarkup&quot;: &quot;&lt;div class=\\&quot;absolute top-1/2 end-3 -translate-y-1/2\\&quot;&gt;&lt;svg class=\\&quot;shrink-0 size-3.5 text-gray-500 dark:text-neutral-500 \\&quot; xmlns=\\&quot;http://www.w3.org/2000/svg\\&quot; width=\\&quot;24\\&quot; height=\\&quot;24\\&quot; viewBox=\\&quot;0 0 24 24\\&quot; fill=\\&quot;none\\&quot; stroke=\\&quot;currentColor\\&quot; stroke-width=\\&quot;2\\&quot; stroke-linecap=\\&quot;round\\&quot; stroke-linejoin=\\&quot;round\\&quot;&gt;&lt;path d=\\&quot;m7 15 5 5 5-5\\&quot;/&gt;&lt;path d=\\&quot;m7 9 5-5 5 5\\&quot;/&gt;&lt;/svg&gt;&lt;/div&gt;&quot;
      }" class="hidden"><option value=""></option><option>Option 1</option><option>Option 2</option><option>Option 3</option></select></div>"
`;
